packages/neuron-wallet/src/services/asset-account-service.ts

```diff
@@ -0,0 +1,323 @@
+      .onConflict(`("tokenID", "blake160") DO NOTHING`)
+      .execute()
+  }
+
+  public static async checkAndDeleteWhenFork(startBlockNumber: string, anyoneCanPayLockHashes: string[]) {
+    const startBlockNumberInt = BigInt(startBlockNumber)
+
+    const assetAccountInfo = new AssetAccountInfo()
+    const outputs = await getConnection()
+      .getRepository(OutputEntity)
+      .createQueryBuilder('output')
+      .leftJoinAndSelect('output.transaction', 'tx')
+      .select('output.lockHash', 'lockHash')
+      .addSelect('output.lockArgs', 'lockArgs')
+      .addSelect('output.typeArgs', 'typeArgs')
+      .addSelect('group_concat(tx.status)', 'txStatusArray')
+      .addSelect('group_concat(tx.blockNumber)', 'blockNumberArray')
+      .where(`output.lockHash IN (:...lockHashes) AND (output.typeCodeHash IS NULL OR (output.typeCodeHash = :sudtCodeHash AND output.typeHashType = :sudtHashType))`, {
+        lockHashes: anyoneCanPayLockHashes,
+        sudtCodeHash: assetAccountInfo.infos.sudt.codeHash,
+        sudtHashType: assetAccountInfo.infos.sudt.hashType,
+      })
+      .groupBy('output.lockHash')
+      .addGroupBy('output.typeArgs')
+      .getRawMany()
+
+    const result = outputs
+      .filter(o => {
+        const status = (o.txStatusArray as string).split(',').map(a => a.trim())
+        const blockNumbers: bigint[] = (o.blockNumberArray as string).split(',').map(a => BigInt(a.trim())).sort()
+        if (
+          blockNumbers[blockNumbers.length - 1] >= startBlockNumberInt &&
+          !(blockNumbers[0] < startBlockNumberInt) &&
+          !status.includes(TransactionStatus.Pending)
+        ) {
+          return o
+        }
+        return undefined
+      })
+
+    for (const output of result) {
+      await getConnection()
+        .createQueryBuilder()
+        .delete()
+        .from(AssetAccountEntity)
+        .where("tokenID = :tokenID AND blake160 = :blake160", {
+          tokenID: output.typeArgs || 'CKBytes',
+          blake160: output.lockArgs,
+        })
+        .execute()
+    }
+  }
+
+  private static async blake160sOfAssetAccounts(): Promise<string[]> {
+    const assetAccounts = await getConnection()
+      .getRepository(AssetAccountEntity)
+      .createQueryBuilder('aa')
+      .select('aa.blake160', 'blake160')
+      .getRawMany()
+
+    return assetAccounts.map(aa => aa.blake160)
+  }
+
+  public static async sendTx(walletID: string, assetAccount: AssetAccount, tx: Transaction, password: string): Promise<string> {
+    // 1. save AssetAccount
+    const connection = getConnection()
+    const entity = AssetAccountEntity.fromModel(assetAccount)
+    const savedEntity = await connection.manager.save([entity.sudtTokenInfo, entity])
+
+    // 2. send tx
+    // if failed, remove saved entity
+    let txHash: string | undefined
+    try {
+      txHash = await new TransactionSender().sendTx(walletID, tx, password)
+    } catch (err) {
+      await connection.manager.remove(savedEntity)
```

Comment:
I think it would be valuable to keep this as an failed record instead of deleting them when the transaction is failed.

Having these failed transaction in DB would be helpful for the users to review the whole transaction history and also helpful when it comes to technical debugging.

@classicalliu @Keith-CY @kellyshang Let me know what you think :)